datastream-server/src/main/java/com/linkedin/datastream/server/Coordinator.java

```diff
@@ -223,6 +226,14 @@ public class Coordinator implements ZkAdapter.ZkAdapterListener, MetricsAware {
   private Future<?> _leaderDoAssignmentScheduledFuture = null;
   private volatile boolean _zkSessionExpired = false;
 
+  // Cache all the throughput violating topics per datastream
+  private final Map<String, Set<String>> _throughputViolatingTopicsMap = new HashMap<>();
+
+  // As the _throughputViolatingTopicsMap may be updated by watcher thread and read by multiple task threads concurrently.
+  private final ReadWriteLock _throughputViolatingTopicsMapReadWriteLock = new ReentrantReadWriteLock();
+  private final Lock _throughputViolatingTopicsMapWriteLock = _throughputViolatingTopicsMapReadWriteLock.writeLock();
+  private final Lock _throughputViolatingTopicsMapReadLock = _throughputViolatingTopicsMapReadWriteLock.readLock();
+
```

Comment:
From @vmaheshw 
```
Do we need to go to low-level locks for this? Can we not rely on ConcurrentHashMap? 
Violation calculation does not have to be precise in the close second.
```

Reply:
I would prefer not to use the concurrent hash map since while performing "replace-all" to this map, there may be chances that these violating topics' metrics and SLAs get reported. Hence to keep things consistent, I went with this approach.

Reply:
I recommend simplicity, especially since this is not on a critical path.